### PR TITLE
feat: splitpanes migration (broken draft)

### PR DESCRIPTION
**Pull request by hopperelec**
As mentioned in PR https://github.com/LeoDog896/svelte-mosaic/pull/1, a lot of svelte-mosaic's current functionality has already been implemented for [svelte-splitpanes](https://github.com/orefalo/svelte-splitpanes/). This is what I've done so far in my attempt to migrate svelte-mosaic's `Node`s to svelte-splitpanes. As part of this, I have also significantly refactored the `Tree` interface. However, one issue I have encountered is that splitpanes uses % sizing, whereas mosaic currently uses px sizing. I have tried to get around this, both by wrapping panes in a `<div>` with size constraints and by calculating the px values as a percentage of the splitpanes size (this is what is shown in this draft), but I've not managed to get it working.

It appears there is effort going on to implement px sizing in splitpanes (see https://github.com/orefalo/svelte-splitpanes/issues/33). So, we could wait for something to come out of that or we could switch to using % for the time being. Thoughts?

### CODE REVIEW

**Comment by tristan-f-r** [comment]
> As mentioned in PR #1, a lot of svelte-mosaic's current functionality has already been implemented for [svelte-splitpanes](https://github.com/orefalo/svelte-splitpanes/). This is what I've done so far in my attempt to migrate svelte-mosaic's `Node`s to svelte-splitpanes. As part of this, I have also significantly refactored the `Tree` interface. However, one issue I have encountered is that splitpanes uses % sizing, whereas mosaic currently uses px sizing. I have tried to get around this, both by wrapping panes in a `<div>` with size constraints and by calculating the px values as a percentage of the splitpanes size (this is what is shown in this draft), but I've not managed to get it working.
> 
> It appears there is effort going on to implement px sizing in splitpanes (see [orefalo/svelte-splitpanes#33](https://github.com/orefalo/svelte-splitpanes/issues/33)). So, we could wait for something to come out of that or we could switch to using px for the time being. Thoughts?

That compromise is alright - though, hardcoding it to `px` wasn't the best decision on my part either - that should be changed from `number` to `{ unit: 'pixel' | 'percentage', amount: number }` (with `pixel` being unsupported for now)

**Comment by hopperelec** [comment]
I definitely agree it should be a choice! I honestly think `px` is more useful and I don't really understand why splitpanes is so hesitant about it, but I can see why `%` would also be useful

**Comment by hopperelec** [comment]
Just realised the issue I was running into trying to use percentages was nothing to do with the way I was converting the px values- it's a known bug in the latest version of splitpanes (https://github.com/orefalo/svelte-splitpanes/issues/96). So, we could actually use the code I wrote for that first commit, as a sorta polyfill until it's properly implemented in splitpanes, and add on support for `%` too.

**Comment by hopperelec** [comment]
We could also add support for other CSS units (e.g: `em`, `ch`, viewport, `pt`) but, to avoid over-complicating things and to stick to the approach splitpanes appears to be wanting to take, I'll stick with just `px` and `%` for now

**Comment by hopperelec** [comment]
While testing, I've noticed that if the size range does not contain `50%`, the message below is shown to the console and, if the size range is below `50%`, the panes won't fill the container.
> Splitpanes: Could not resize panes correctly due to their constraints.

I'm guessing this is intentional and because splitpanes is expecting you to set the default size using the `size` prop. I think mosaic should provide an option for specifying a default size, but do you think it would be worth making it default to:
- If the size range is below `50%`, set the default size to the upper limit of the size range
- If the size range is above `50%`, set the default size to the lower limit of the size range

**Comment by hopperelec** [comment]
As well as `size`, I think it's also worth adding support for all other props provided by splitpanes, such as `id`, `theme`, `class` and `snapSize`, but I want your input on that first

**Comment by tristan-f-r** [comment]
> As well as `size`, I think it's also worth adding support for all other props provided by splitpanes, such as `id`, `theme`, `class` and `snapSize`, but I want your input on that first

That sounds fine - the API should be dependent on `splitpanes`, not independent of it.

**Comment by hopperelec** [comment]
The initial size isn't working correctly. If the initial size is set, even if to 50%, for some reason it starts at 0. I'm suspecting this is another bug in splitpanes, especially since it doesn't even show the "Could not resize panes correctly due to their constraints." warning when this happens, but I can't find any reports about this one.

**Comment by tristan-f-r** [comment]
> The initial size isn't working correctly. If the initial size is set, even if to 50%, for some reason it starts at 0. I'm suspecting this is another bug in splitpanes, especially since it doesn't even show the "Could not resize panes correctly due to their constraints." warning when this happens, but I can't find any reports about this one.

It had to do with:
- Incorrect units passed (svelte-splitpanes is looking for a number from 0-100, the functions switched between 0-1 and 0-100)
- Siblings needed balanced percentages - I've edited it so they know their sibling's desired initial number.

Edit: I just realized my changes did not fix the issue.

**Comment by hopperelec** [comment]
> Incorrect units passed (svelte-splitpanes is looking for a number from 0-100, the functions switched between 0-1 and 0-100)

I'm confused, because I checked using svelte-devtools that it was definitely setting it to values between 0-100, and your changes look to be now making it between 0-1.

> Siblings needed balanced percentages - I've edited it so they know their sibling's desired initial number.

svelte-splitpanes is supposed to handle this automatically, to my understanding

**Comment by tristan-f-r** [comment]
> svelte-splitpanes is supposed to handle this automatically, to my understanding

It seems to be the case. I just fixed it with the latest commit, and I've remove the sibling handling, though the documentation seems to be outdated about that.

The actual issue was `containerSizePx` not being set, making the initial a significantly tiny number.

**Comment by hopperelec** [comment]
The current code is giving me the
> Splitpanes: Could not resize panes correctly due to their constraints.

warning again, and even after triggering a re-size it is not displaying correctly. I think it's to do with the way you changed the 0-1 and 0-100 handling, but I'm finding this logic hard to follow

**Comment by hopperelec** [comment]
> The actual issue was `containerSizePx` not being set, making the initial a significantly tiny number.

I've reverted it back to the old 0-100 logic locally to test, and it just seems to still be doing the same as before where it starts with 0 size. I've also tried specifically dividing the initial size by 100 to ensure splitpanes wasn't just handling percentages for that one differently, and that hasn't fixed it.

**Comment by tristan-f-r** [comment]
> > The actual issue was `containerSizePx` not being set, making the initial a significantly tiny number.
> 
> I've reverted it back to the old 0-100 logic locally to test, and it just seems to still be doing the same as before where it starts with 0 size. I've also tried specifically dividing the initial size by 100 to ensure splitpanes wasn't just handling percentages for that one differently, and that hasn't fixed it.

Odd - I'm not experiencing that issue locally.

**Comment by hopperelec** [comment]
> The initial size isn't working correctly. If the initial size is set, even if to 50%, for some reason it starts at 0

Just for the record: there is a discrepancy between the result in Chromium and Firefox. In Chromium (which I was using), it starts at 0. But in Firefox, it actually starts slightly higher than the initial size requested. We have no idea what could be causing this discrepancy, nor why it is not working in the first place, but I'm confident that it's a bug in splitpanes and not in mosaic.
